File: R/render_stimuli.R
# =============================================================================
# render_stimuli.R — Batch renderer
# =============================================================================
# Takes a trial data frame (or any subset) and renders each row to an image.
# Has no opinions about experimental design — renders whatever it receives.
# =============================================================================

source("Ebbinghaus/R/draw_trial.R")

#' Render stimulus images for a trial table
#'
#' @param trials A data frame with the trials table schema (or a subset).
#' @param verbose Logical: print progress messages. Default TRUE.
#'
#' @return Invisibly returns the input trials data frame. Side effect: writes
#'   image files to the paths specified in each row's `file_path` column.
render_stimuli <- function(trials, verbose = TRUE) {

  stopifnot(is.data.frame(trials), nrow(trials) > 0)

  # Ensure output directories exist
  dirs <- unique(dirname(trials$file_path))
  for (d in dirs) {
    if (!dir.exists(d)) dir.create(d, recursive = TRUE)
  }

  n <- nrow(trials)
  errors <- character(0)

  for (i in seq_len(n)) {
    row <- trials[i, ]
    fpath <- row$file_path
    fmt <- row$file_format
    w <- row$canvas_width
    h <- row$canvas_height

    if (verbose && (i %% 10 == 0 || i == 1 || i == n)) {
      message(sprintf("[%d/%d] Rendering trial %d -> %s", i, n,
                      row$trial_id, fpath))
    }

    tryCatch({
      p <- draw_trial(row)

      if (fmt == "png") {
        ggsave(fpath, plot = p, width = w / 96, height = h / 96,
               dpi = 96, units = "in", bg = row$background_color)
      } else if (fmt == "svg") {
        ggsave(fpath, plot = p, width = w / 96, height = h / 96,
               units = "in", bg = row$background_color, device = "svg")
      } else if (fmt == "webp") {
        # webp via ragg if available, otherwise fall back to png
        if (requireNamespace("ragg", quietly = TRUE)) {
          ggsave(fpath, plot = p, width = w / 96, height = h / 96,
                 dpi = 96, units = "in", bg = row$background_color,
                 device = ragg::agg_png)
          # Rename .png to .webp — or use a proper webp device if available
          # For now, fall back to png with webp extension as placeholder
          warning("webp support is placeholder; saving as png with .webp extension")
        } else {
          warning("ragg not available; saving webp as png with .webp extension")
          ggsave(fpath, plot = p, width = w / 96, height = h / 96,
                 dpi = 96, units = "in", bg = row$background_color)
        }
      } else {
        stop("Unsupported file format: ", fmt)
      }

      if (!file.exists(fpath)) {
        errors <- c(errors, paste0("Trial ", row$trial_id, ": file not created"))
      }
    },
    error = function(e) {
      msg <- paste0("Trial ", row$trial_id, ": ", conditionMessage(e))
      errors <<- c(errors, msg)
      if (verbose) message("  ERROR: ", msg)
    })
  }

  if (length(errors) > 0) {
    warning(length(errors), " errors during rendering:\n",
            paste("  -", errors, collapse = "\n"))
  } else if (verbose) {
    message("All ", n, " stimuli rendered successfully.")
  }

  invisible(trials)
}

File: py/src/render_stimuli.py
"""
render_stimuli.py — Batch renderer

Takes a trial DataFrame (or any subset) and renders each row to an image.
"""

import os
import pandas as pd
import matplotlib.pyplot as plt

from src.draw_trial import draw_trial


def render_stimuli(trials: pd.DataFrame, verbose: bool = True) -> pd.DataFrame:
    """Render stimulus images for a trial table.

    Args:
        trials: DataFrame with the trials table schema (or a subset).
        verbose: Print progress messages.

    Returns:
        The input trials DataFrame (unchanged). Side effect: writes
        image files to paths specified in each row's `file_path` column.
    """
    assert isinstance(trials, pd.DataFrame) and len(trials) > 0

    # Ensure output directories exist
    dirs = trials["file_path"].apply(os.path.dirname).unique()
    for d in dirs:
        os.makedirs(d, exist_ok=True)

    n = len(trials)
    errors = []

    for i in range(n):
        row = trials.iloc[i]
        fpath = row["file_path"]
        fmt = row["file_format"]
        w = row["canvas_width"]
        h = row["canvas_height"]

        if verbose and (i % 10 == 0 or i == 0 or i == n - 1):
            print(f"[{i + 1}/{n}] Rendering trial {row['trial_id']} -> {fpath}")

        try:
            fig = draw_trial(row.to_frame().T)

            if fmt == "png":
                fig.savefig(fpath, dpi=96, facecolor=fig.get_facecolor())
            elif fmt == "svg":
                fig.savefig(fpath, format="svg", facecolor=fig.get_facecolor())
            elif fmt == "webp":
                fig.savefig(fpath, format="webp", dpi=96, facecolor=fig.get_facecolor())
            else:
                raise ValueError(f"Unsupported file format: {fmt}")

            plt.close(fig)

            if not os.path.exists(fpath):
                errors.append(f"Trial {row['trial_id']}: file not created")

        except Exception as e:
            errors.append(f"Trial {row['trial_id']}: {e}")
            if verbose:
                print(f"  ERROR: {e}")

    if errors:
        print(f"WARNING: {len(errors)} errors during rendering:")
        for err in errors:
            print(f"  - {err}")
    elif verbose:
        print(f"All {n} stimuli rendered successfully.")

    return trials
